### PR TITLE
Port `lithium_ion_battery.m` to Python

### LINKED ISSUE #1259: Lithium-ion battery example (and unit tests) missing from Python
**Problem description**

While the `lithium_ion_battery.m` example is available in MATLAB, it is missing from Python, meaning that corresponding code portions are not covered by CI, see discussion in #1256 (specifically, the only instances where `lithium_ion_battery.cti/yaml` are used are the MATLAB example, plus a conversion test that is not designed to probe any meaningful features).

While this isn't a bug specifically, I view it as a critical omission that hopefully will get fixed prior to the release of 2.6. At the minimum, the MATLAB test should be converted to a unit test similar to `test_kinetics.py::TestSofcKinetics`.

**System information**

- Cantera version: 2.6.0b2
- OS: any
- Python/MATLAB/other software versions: any

### PR DESCRIPTION
**Pull request by ischoegl**
**Changes proposed in this pull request**

Port `lithium_ion_battery.m` to Python. Ultimately, this was the most efficient resolution.

**If applicable, fill in the issue number this pull request is fixing**

Closes #1259

**If applicable, provide an example illustrating new features this pull request is introducing**

Results ... @decaluwe, please confirm that this is consistent, as I currently don't have a MATLAB version installed.

lithium_ion_battery

If the CTI input is used, the output is zero (see #1256). I will add a unit test once #1256 is merged.

**Checklist**

- [x] The pull request includes a clear description of this code change
- [x] Commit messages have short titles and reference relevant issues
- [x] Build passes (`scons build` & `scons test`) and unit tests address code coverage
- [x] Style & formatting of contributed code follows [contributing guidelines](https://github.com/Cantera/cantera/blob/main/CONTRIBUTING.md)
- [x] The pull request is ready for review

### CODE REVIEW

**Comment by speth** [comment]
Here's the MATLAB version of that graph -- looks good to me:
image

**Comment by ischoegl** [comment]
This is ready for review … @decaluwe please feel free to push suggestions directly to this branch.

File: interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+    return - r * ct.faraday * area_cathode
+
+
+# Calculate cell voltage, separately for each entry of the input vectors
+V_cell = np.zeros_like(soc)
+phi_l_anode = 0
+phi_s_cathode = 0
+for i in range(samples):
+    # Set anode electrode potential to 0
+    phi_s_anode = 0;
```
Comment (speth) [inline interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py]:
```suggestion
    phi_s_anode = 0
```

File: interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py [hunk 2]
```diff
@@ -0,0 +1,172 @@
...
+
+    # Calculate anode electrolyte potential
+    phi_l_anode = newton_solve(
+        lambda E: anode_current(phi_s_anode, E, X_Li_anode[i]),
+        phi_l_anode, C=current)
+
+    # Calculate cathode electrolyte potential
+    phi_l_cathode = phi_l_anode + current * R_electrolyte;
```
Comment (speth) [inline interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py]:
```suggestion
    phi_l_cathode = phi_l_anode + current * R_electrolyte
```

File: interfaces/cython/cantera/test/test_kinetics.py [hunk 3]
```diff
@@ -916,6 +916,106 @@ class TestSofcKinetics3(TestSofcKinetics):
...
+            # Calculate the current. Should be negative for cell discharge.
+            return - ct.faraday * cathode_int.net_rates_of_progress * area_cathode
+
+        # Calculate cell voltage, separately for each entry of the input vectors
+        data = []
+        phi_l_anode = 0
+        phi_s_cathode = 0
+        for i in range(samples):
+            # Calculate anode electrolyte potential
+            phi_s_anode = 0;
```
Comment (speth) [inline interfaces/cython/cantera/test/test_kinetics.py]:
```suggestion
            phi_s_anode = 0
```

File: interfaces/cython/cantera/test/test_kinetics.py [hunk 4]
```diff
@@ -916,6 +916,106 @@ class TestSofcKinetics3(TestSofcKinetics):
...
+            phi_l_anode = newton_solve(
+                lambda E: anode_current(phi_s_anode, E, X_Li_anode[i]),
+                phi_l_anode, C=current)
+
+            # Calculate cathode electrode potential
+            phi_l_cathode = phi_l_anode + current * R_electrolyte;
```
Comment (speth) [inline interfaces/cython/cantera/test/test_kinetics.py]:
```suggestion
            phi_l_cathode = phi_l_anode + current * R_electrolyte
```

File: interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py [hunk 5]
```diff
@@ -0,0 +1,172 @@
...
+
+
+# This function returns the Cantera calculated anode current (in A)
+def anode_current(phi_s, phi_l, X_Li_anode):
+    """
+    Current from the anode as a function of anode potential relative to
+    electrolyte.
+    """
+    # Set the active material mole fraction
+    anode.X = f"Li[anode]:{X_Li_anode},V[anode]:{1 - X_Li_anode}"
```
Comment (speth) [inline interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py]: I know this is borrowed from the Jupyter example, but I think it's better to write this as
```suggestion
    anode.X = {"Li[anode]": X_Li_anode, "V[anode]": 1 - X_Li_anode}
```
(also in cathode_current, and in the test)

**Review by speth — DISMISSED** [review]
Thanks, @ischoegl, I think this is good. I had just a couple of very minor formatting comments.

File: interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py (hunk 5 above)
Comment (ischoegl) [reply interfaces/cython/cantera/examples/surface_chemistry/lithium_ion_battery.py]: :tada: ... this was transcribed from MATLAB, but you're absolutely right!

**Comment by ischoegl** [comment]
@speth ... thanks for the quick review! Re-approval is needed as I squashed the fixes ...

**Review by speth — APPROVED** [review]

**Pull request merged**